From 501d968b7b25ae3442f337a739edc0a7108e8eed Mon Sep 17 00:00:00 2001
From: Arnaud Quette <arnaud.quette@free.fr>
Date: Thu, 23 Jun 2016 13:20:04 +0200
Subject: [PATCH 1/2] netxml-ups: fix a segmentation fault

netxml_get_page() was not checking if page is NULL, hence resulting in a
segmentation fault with Eaton ePDU, which use the unsupported version 4 of the
XML/PDC protocol. Bump the driver version to 0.41
---
 drivers/netxml-ups.c | 27 ++++++++++++++-------------
 1 file changed, 14 insertions(+), 13 deletions(-)

diff --git a/drivers/netxml-ups.c b/drivers/netxml-ups.c
index 2568ff0..0b237bc 100644
--- a/drivers/netxml-ups.c
+++ b/drivers/netxml-ups.c
@@ -40,7 +40,7 @@
 #include <ne_socket.h>
 
 #define DRIVER_NAME	"network XML UPS"
-#define DRIVER_VERSION	"0.40"
+#define DRIVER_VERSION	"0.41"
 
 /** *_OBJECT query multi-part body boundary */
 #define FORM_POST_BOUNDARY "NUT-NETXML-UPS-OBJECTS"
@@ -656,27 +656,28 @@ void upsdrv_cleanup(void)
 
 static int netxml_get_page(const char *page)
 {
-	int		ret;
+	int		ret = NE_ERROR;
 	ne_request	*request;
 	ne_xml_parser	*parser;
 
-	upsdebugx(2, "%s: %s", __func__, page);
-
-	request = ne_request_create(session, "GET", page);
+	upsdebugx(2, "%s: %s", __func__, (page != NULL)?page:"(null)");
 
-	parser = ne_xml_create();
+	if (page != NULL) {
+		request = ne_request_create(session, "GET", page);
 
-	ne_xml_push_handler(parser, subdriver->startelm_cb, subdriver->cdata_cb, subdriver->endelm_cb, NULL);
+		parser = ne_xml_create();
 
-	ret = netxml_dispatch_request(request, parser);
+		ne_xml_push_handler(parser, subdriver->startelm_cb, subdriver->cdata_cb, subdriver->endelm_cb, NULL);
 
-	if (ret) {
-		upsdebugx(2, "%s: %s", __func__, ne_get_error(session));
-	}
+		ret = netxml_dispatch_request(request, parser);
 
-	ne_xml_destroy(parser);
-	ne_request_destroy(request);
+		if (ret) {
+			upsdebugx(2, "%s: %s", __func__, ne_get_error(session));
+		}
 
+		ne_xml_destroy(parser);
+		ne_request_destroy(request);
+	}
 	return ret;
 }
 
-- 
2.1.4


From 77625a29e47ac2a8261ba62db998cf160b29f3e3 Mon Sep 17 00:00:00 2001
From: Arnaud Quette <arnaud.quette@free.fr>
Date: Thu, 23 Jun 2016 13:22:30 +0200
Subject: [PATCH 2/2] netxml-ups: detect the unsupported XML v4 protocol

The XML/PDC protocol had an update with version 4, for ePDU. However, this
version is not supported by netxml-ups. We now detect this version, and
explicitly exit with an "XML v4 protocol is not supported!" message
---
 drivers/mge-xml.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/mge-xml.c b/drivers/mge-xml.c
index 8ca4dd3..172b1d8 100644
--- a/drivers/mge-xml.c
+++ b/drivers/mge-xml.c
@@ -31,7 +31,7 @@
 #include "netxml-ups.h"
 #include "mge-xml.h"
 
-#define MGE_XML_VERSION		"MGEXML/0.25"
+#define MGE_XML_VERSION		"MGEXML/0.26"
 #define MGE_XML_INITUPS		"/"
 #define MGE_XML_INITINFO	"/mgeups/product.xml /product.xml /ws/product.xml"
 
@@ -1085,6 +1085,7 @@ static int mge_xml_startelm_cb(void *userdata, int parent, const char *nspace, c
 			/* name="Network Management Card" type="Mosaic M" version="BA" */
 			/* name="Network Management Card" type="Transverse" version="GB (SN 49EH29101)" */
 			/* name="Monitored ePDU" type="Monitored ePDU" version="Version Upgrade" */
+			/* name="PDU Network Management Card" type="SCOB" version="02.00.0036" signature="34008876" protocol="XML.V4" */
 			int	i;
 			for (i = 0; atts[i] && atts[i+1]; i += 2) {
 				if (!strcasecmp(atts[i], "name")) {
@@ -1108,6 +1109,13 @@ static int mge_xml_startelm_cb(void *userdata, int parent, const char *nspace, c
 					}
 					dstate_setinfo("ups.firmware.aux", "%s", val);
 				}
+				/* netxml-ups currently only supports XML version 3 (for UPS),
+				 * and not version 4 (for UPS and PDU)! */
+				if (!strcasecmp(atts[i], "protocol")) {
+					if (!strcasecmp(atts[i+1], "XML.V4")) {
+						fatalx(EXIT_FAILURE, "XML v4 protocol is not supported!");
+					}
+				}
 			}
 			state = PRODUCT_INFO;
 			break;
-- 
2.1.4

